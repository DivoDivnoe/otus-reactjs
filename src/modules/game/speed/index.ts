import reducer from './speed';

export { ActionCreator, SpeedType, SpeedValue, SpeedValueType } from './speed';
export { getSpeed } from './selectors';

export default reducer;
