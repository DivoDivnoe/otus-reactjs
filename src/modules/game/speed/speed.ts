import { createSlice, PayloadAction } from '@reduxjs/toolkit';

export enum SpeedType {
  FAST = 'fast',
  MEDIUM = 'medium',
  SLOW = 'slow',
}

export interface SpeedValueType {
  [SpeedType.FAST]: number;
  [SpeedType.MEDIUM]: number;
  [SpeedType.SLOW]: number;
}

// timeout interval ms
export const SpeedValue: SpeedValueType = {
  [SpeedType.FAST]: 30,
  [SpeedType.MEDIUM]: 150,
  [SpeedType.SLOW]: 500,
};

const speedSlice = createSlice({
  name: 'speed',
  initialState: SpeedType.MEDIUM,
  reducers: {
    setSpeed: (_state, action: PayloadAction<SpeedType>) => action.payload,
  },
});

export const ActionCreator = speedSlice.actions;
export default speedSlice.reducer;
