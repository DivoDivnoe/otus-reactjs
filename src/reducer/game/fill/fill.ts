import { FillType } from '@/constants';
import { gameProps } from '@/configs';
import { createSlice, PayloadAction } from '@reduxjs/toolkit';

const initialState: FillType = gameProps.fill;

const fillSlice = createSlice({
  name: 'fill',
  initialState,
  reducers: {
    setFill: (_state, action: PayloadAction<FillType>) => action.payload,
  },
});

export const ActionCreator = fillSlice.actions;
export default fillSlice.reducer;
