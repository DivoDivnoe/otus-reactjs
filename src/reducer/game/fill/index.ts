import reducer from './fill';

export {
  ActionCreator,
  FillType,
  BoardFillPercentage,
  BoardFillPercentageType,
} from './fill';
export { getFill } from './selectors';

export default reducer;
