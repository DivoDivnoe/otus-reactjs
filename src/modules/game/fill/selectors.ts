import { NAME_SPACE } from '@/modules/game/nameSpace';
import { State } from '@/reducer';
import { FillType } from '.';

export const getFill = (state: State): FillType => state[NAME_SPACE].fill;
