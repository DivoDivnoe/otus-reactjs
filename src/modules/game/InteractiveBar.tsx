import React, { FC, useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Bar } from '@/components/Bar';
import { gameOptions } from '@/configs';
import { State } from '@/reducer';
import { BoardSize, getSize, ActionCreator as SizeActionCreator } from './size';
import {
  SpeedType,
  getSpeed,
  ActionCreator as SpeedActionCreator,
} from './speed';
import { FillType, getFill, ActionCreator as FillActionCreator } from './fill';
import {
  getIsPlaying,
  ActionCreator as IsPlayingActionCreator,
} from './isPlaying';
import {
  Model,
  ActionCreator as ModelActionCreator,
  getZeroMatrix,
} from './model';

export const InteractiveBar: FC = () => {
  const dispatch = useDispatch();

  const { sizes, speedTypes, fillTypes } = gameOptions;

  const size = useSelector<State, BoardSize>(getSize);
  const speed = useSelector<State, SpeedType>(getSpeed);
  const fill = useSelector<State, FillType>(getFill);
  const isPlaying = useSelector<State, boolean>(getIsPlaying);
  const zeroMatrix = useSelector<State, Model>(getZeroMatrix);

  const changeSizeHandler = useCallback(
    (size: BoardSize) => {
      dispatch(SizeActionCreator.setSize(size));
    },
    // Stryker disable next-line ArrayDeclaration
    []
  );

  const changeSpeedHandler = useCallback(
    (speedType: SpeedType) => {
      dispatch(SpeedActionCreator.setSpeed(speedType));
    },
    // Stryker disable next-line ArrayDeclaration
    []
  );

  const changeFillType = useCallback(
    (fill: FillType) => {
      dispatch(FillActionCreator.setFill(fill));
    },
    // Stryker disable next-line ArrayDeclaration
    []
  );

  const play = useCallback(
    () => {
      dispatch(IsPlayingActionCreator.startPlaying());
    },
    // Stryker disable next-line ArrayDeclaration
    []
  );
  const pause = useCallback(
    () => {
      dispatch(IsPlayingActionCreator.stopPlaying());
    },
    // Stryker disable next-line ArrayDeclaration
    []
  );

  const clear = useCallback(() => {
    dispatch(ModelActionCreator.setModel(zeroMatrix));
  }, [size]);

  return (
    <Bar
      size={size}
      speed={speed}
      fill={fill}
      sizes={sizes}
      speedTypes={speedTypes}
      fillTypes={fillTypes}
      changeSizeHandler={changeSizeHandler}
      changeSpeedHandler={changeSpeedHandler}
      changeFillType={changeFillType}
      play={play}
      pause={pause}
      clear={clear}
      isPlaying={isPlaying}
    />
  );
};
