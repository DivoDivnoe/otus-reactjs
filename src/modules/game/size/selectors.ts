import { NAME_SPACE } from '@/modules/game/nameSpace';
import { State } from '@/reducer';
import { BoardSize } from '.';

export const getSize = (state: State): BoardSize => state[NAME_SPACE].size;
