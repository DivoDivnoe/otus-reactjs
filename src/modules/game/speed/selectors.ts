import { NAME_SPACE } from '@/modules/game/nameSpace';
import { State } from '@/reducer';
import { SpeedType } from '.';

export const getSpeed = (state: State): SpeedType => state[NAME_SPACE].speed;
