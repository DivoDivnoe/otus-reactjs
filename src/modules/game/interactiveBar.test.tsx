import React from 'react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useSelector, useDispatch } from 'react-redux';
import { InteractiveBar } from './InteractiveBar';
import { BoardSize } from './size';
import { SpeedType } from './speed';
import { FillType } from './fill';
import reducer from '@/reducer';

jest.mock('react-redux', () => ({
  __esModule: true,
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

describe('InteractiveBar test cases', () => {
  it('InteractiveBar check', () => {
    const mockDispatch = jest.fn();

    (useDispatch as jest.Mock).mockReturnValue(mockDispatch);
    (useSelector as jest.Mock).mockImplementation((callback) => {
      return callback(store.getState());
    });

    const store = configureStore({ reducer });

    const { asFragment } = render(
      <Provider store={store}>
        <InteractiveBar />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();

    userEvent.click(screen.getByRole('button', { name: 'Size: 50x30' }));
    expect(mockDispatch).toHaveBeenCalledWith({
      type: 'size/setSize',
      payload: BoardSize.SMALL,
    });

    userEvent.click(screen.getByRole('button', { name: '30%' }));
    expect(mockDispatch).toHaveBeenCalledWith({
      type: 'fill/setFill',
      payload: FillType.HIGH,
    });

    userEvent.click(screen.getByRole('button', { name: 'fast' }));
    expect(mockDispatch).toHaveBeenCalledWith({
      type: 'speed/setSpeed',
      payload: SpeedType.FAST,
    });
  });
});
