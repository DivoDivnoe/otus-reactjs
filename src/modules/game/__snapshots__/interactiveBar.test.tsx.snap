// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InteractiveBar test cases InteractiveBar check 1`] = `
<DocumentFragment>
  <div
    class="css-oyac82"
  >
    <div
      class="css-1aljmas"
    >
      <span
        class="css-c5nu7q"
      >
        Board Size: 
      </span>
      <div
        class="css-lue796"
      >
        <button
          class="css-som8ht"
        >
          Size: 50x30
        </button>
        <button
          class="css-1u5m4u0"
        >
          Size: 70x50
        </button>
        <button
          class="css-som8ht"
        >
          Size: 100x80
        </button>
      </div>
    </div>
    <div
      class="css-1aljmas"
    >
      <span
        class="css-c5nu7q"
      >
        Sim Speed: 
      </span>
      <div
        class="css-lue796"
      >
        <button
          class="css-som8ht"
        >
          slow
        </button>
        <button
          class="css-1u5m4u0"
        >
          medium
        </button>
        <button
          class="css-som8ht"
        >
          fast
        </button>
      </div>
    </div>
    <div
      class="css-1aljmas"
    >
      <span
        class="css-c5nu7q"
      >
        Fill: 
      </span>
      <div
        class="css-lue796"
      >
        <button
          class="css-som8ht"
        >
          10%
        </button>
        <button
          class="css-nm0pqm"
        >
          20%
        </button>
        <button
          class="css-som8ht"
        >
          30%
        </button>
      </div>
    </div>
    <div
      class="css-1aljmas"
    >
      <span
        class="css-c5nu7q"
      >
        Controls: 
      </span>
      <div
        class="css-lue796"
      >
        <button
          class="css-som8ht"
        >
          play
        </button>
        <button
          class="css-1u5m4u0"
        >
          pause
        </button>
        <button
          class="css-som8ht"
        >
          clear
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
